include/nil/crypto3/math/multithreading/thread_pool.hpp

```diff
@@ -0,0 +1,117 @@
+//---------------------------------------------------------------------------//
+// Copyright (c) 2023 Martun Karapetyan <martun@nil.foundation>
+//
+// MIT License
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in all
+// copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+//---------------------------------------------------------------------------//
+
+#ifndef CRYPTO3_THREAD_POOL_HPP
+#define CRYPTO3_THREAD_POOL_HPP
+
+#include <boost/asio/thread_pool.hpp>
+#include <boost/asio/post.hpp>
+
+#include <functional>
+#include <future>
+#include <thread>
+#include <limits>
+#include <memory>
+#include <stdexcept>
+
+
+namespace nil {
+    namespace crypto3 {
+
+        class ThreadPool {
+        public:
+
+            static ThreadPool& get_instance(std::size_t pool_number, std::size_t pool_size = std::thread::hardware_concurrency()) {
+                static ThreadPool instance0(0, pool_size);
+                static ThreadPool instance1(1, pool_size);
```

Comment:
Both thread pools get initialized simultaneously upon first execution of `get_instance`, but we may only ever need one of them. Is that a problem?

Reply:
I was thinking about creating 2 separate functions, but then I was afraid that the function for instance1 can be simultaneously called from the threads of the intance0, and I'm not sure how safe it is during the initialization of instance1.

In practice (except the tests), 2 pools will always be used.